reactioncommerce/reaction
# GraphQL: Replace "cart/setAnonymousUserEmail" Meteor method with setEmailOnAnonymousCart mutation

Resolves #4563 
Impact: **breaking**  
Type: **feature**

## Changes
- Creates a new `setEmailOnAnonymousCart` GraphQL mutation. Same behavior as the "cart/setAnonymousUserEmail" Meteor method
- The "cart/setAnonymousUserEmail" Meteor method is removed
- The place in checkout where the "cart/setAnonymousUserEmail" Meteor method was called now calls the `setEmailOnAnonymousCart` GraphQL mutation using the Apollo client.
- Updated the `React` Blaze template to include the Apollo client provider.
- Added a new Meteor method, "getOpaqueIdFromInternalId", that can be used to transform IDs from internal to opaque from client code. This is done as a Meteor method because once everything is converted to use GraphQL data instead of Meteor publications, this will no longer be necessary. It is a temporary bridge between older Meteor code and newer code that uses GraphQL.

## Breaking changes
The "cart/setAnonymousUserEmail" Meteor method is removed. This does not break any core behavior, but it might require updates to any custom plugins.

## Testing
- Add items to a cart as an anonymous user (not logged in), and then check out. Choose to check out as guest, and then verify that you can successfully enter an email address and the checkout advances to the next step. This email address form is the only UX change.
- Verify that you can use GraphQL to set the cart email address.

```graphql
mutation SetEmailOnAnonymousCart($input: SetEmailOnAnonymousCartInput!) {
  setEmailOnAnonymousCart(input: $input) {
    cart {
       _id
       email
    }
  }
}
```

Variables:

```js
{ input: { cartId, email, token } }
```
